boa_gc/src/pointers/gc.rs

```diff
@@ -15,11 +15,10 @@ use std::{
     rc::Rc,
 };
 
-use super::rootable::Rootable;
-
 /// A garbage-collected pointer type over an immutable value.
 pub struct Gc<T: Trace + ?Sized + 'static> {
-    pub(crate) inner_ptr: Cell<Rootable<GcBox<T>>>,
+    pub(crate) inner_ptr: Cell<NonNull<GcBox<T>>>,
+    pub(crate) handle: Rc<GcHandle>,
     pub(crate) marker: PhantomData<Rc<T>>,
 }
```

Comment:
Hmmm, can we add another layer of indirection to this?
We were deliberately ensuring this type is pointer sized, otherwise space optimizations such as #1830 will become impossible.

Alternatively, if equal `Gc`s share the same counter, isn't it possible to put the counter inside `GcBox` instead?

Reply:
+1 for @jedel1043's suggestion

Also we should be able to remove the `Cell` from `Gc<T>` internal pointer, since we don't have the old way of rooting, which should make `Option<Gc<T>>` pointer size (instead of two pointer width).

Reply:
Updated to put the counters in GcBox/EphemeronBox. It's a bit slower than the previous one. I update the benchmark result in the description.